File: julia/d3p2.jl
input = raw"xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))"

chunks = split(input, "don't()")
valid = [chunks[1]]
for i in 2:length(chunks)
    if contains(chunks[i], "do()")
        push!(valid, replace(chunks[i], r".*?do\(\)" => ""; count=1))
    end
end

re = r"mul\((\d{1,3}),(\d{1,3})\)"
sum(
    [
        parse(Int64,match.captures[1]) * parse(Int64,match.captures[2]) 
        for match in eachmatch(re, join(valid))
    ]
)

File: rust/src/bin/d3p2.rs
const INPUT: &str = "xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))";

use regex::Regex;
fn main() {
    let chunks: Vec<&str> = INPUT.split("don't()").collect();

    let re_replace = Regex::new(r".*?do\(\)").unwrap();
    let mut valid: Vec<String> = chunks
                .iter()
                .filter(|x| x.contains("do()"))
                .map(|y| re_replace.replacen(y,1,"").into_owned())
                .collect();

    valid.push(chunks[0].to_owned());

    let re_mul = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").unwrap();
    let ans: i32 = re_mul.captures_iter(&valid.join(""))
                    .map(
                        |c| c.get(1).unwrap().as_str().parse::<i32>().unwrap()
                            *
                            c.get(2).unwrap().as_str().parse::<i32>().unwrap()
                    )
                    .sum();

    println!("{:?}", ans);
}